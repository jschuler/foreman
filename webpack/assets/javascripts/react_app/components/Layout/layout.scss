@import '../../common/colors.scss';

.pf-c-page.foreman-theme {
  .pf-c-page__header {
    .pf-c-data-toolbar {
      background-color: transparent;
    }
  }

  .pf-c-page__sidebar.pf-m-dark {
    background-color: $nav-pf-vertical-secondary-active-bg-color;

    .pf-c-nav.pf-m-dark > .pf-c-nav__list {
      .pf-c-nav__item.pf-m-current,
      .pf-c-nav__item > .pf-c-nav__link.pf-m-current {
        background-color: $nav-pf-vertical-active-bg-color;
      }

      .pf-c-nav__item.pf-m-current > .pf-c-nav__link[data-component=pf-nav-expandable] {
        color: $nav-pf-vertical-active-icon-color;
      }

      .pf-c-nav__link {
        color: $nav-pf-vertical-icon-color;
      }

      .pf-c-nav__simple-list .pf-c-nav__link {
        &.pf-m-current,
        &.pf-m-hover,
        &:hover {
          background-color: $nav-pf-vertical-active-bg-color;
          color: $nav-pf-vertical-active-icon-color;
        }
      }
    }
  }
}

#account_menu {
  outline: none;
}

#vertical-spinner {
  margin-top: 15px;
  margin-left: 5px;
}
