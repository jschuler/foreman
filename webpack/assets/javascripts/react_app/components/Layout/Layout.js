import React from 'react';
import {
  Brand,
  Page,
  PageSection,
  PageSectionVariants,
  PageSidebar,
  SkipToContent,
} from '@patternfly/react-core';
// eslint-disable-next-line
import globalBreakpointMd from '@patternfly/react-tokens/dist/js/global_breakpoint_md';
import CustomPageHeader from './components/CustomPageHeader';
import {
  getActive,
  getCurrentPath,
  layoutPropTypes,
  layoutDefaultProps,
} from './LayoutHelper';
import VerticalNav from './components/VerticalNav';
import Toolbar from './components/Toolbar';
import './layout.scss';

class Layout extends React.Component {
  componentDidMount() {
    const {
      items,
      data,
      fetchMenuItems,
      changeLocation,
      currentLocation,
      changeOrganization,
      currentOrganization,
      changeActiveMenu,
      activeMenu,
    } = this.props;
    if (items.length === 0) fetchMenuItems(data);

    const activeURLMenu = getActive(data.menu, getCurrentPath());
    if (activeMenu !== activeURLMenu.title) {
      changeActiveMenu(activeURLMenu);
    }

    if (
      data.taxonomies.locations &&
      !!data.locations.current_location &&
      currentLocation !== data.locations.current_location
    ) {
      const initialLocTitle = data.locations.current_location;
      const initialLocId = data.locations.available_locations.find(
        loc => loc.title === initialLocTitle
      ).id;
      changeLocation({ title: initialLocTitle, id: initialLocId });
    }

    if (
      data.taxonomies.organizations &&
      !!data.orgs.current_org &&
      currentOrganization !== data.orgs.current_org
    ) {
      const initialOrgTitle = data.orgs.current_org;
      const initialOrgId = data.orgs.available_organizations.find(
        org => org.title === initialOrgTitle
      ).id;
      changeOrganization({ title: initialOrgTitle, id: initialOrgId });
    }
  }

  render() {
    const {
      items,
      data,
      isLoading,
      changeActiveMenu,
      changeOrganization,
      changeLocation,
      currentOrganization,
      currentLocation,
      activeMenu,
      children,
      history,
    } = this.props;

    const afterNavToggle = navOpen => {
      const mobileView =
        window.innerWidth < Number.parseInt(globalBreakpointMd.value, 10);
      if (mobileView) {
        // ignore for mobile view
        return;
      }
      // toggles a class in the body tag, so that the main #rails-app-content container can have the appropriate width
      if (navOpen) {
        document.body.classList.add('pf-m-expanded');
      } else {
        document.body.classList.remove('pf-m-expanded');
      }
    };
    const header = (
      <CustomPageHeader
        logo={
          <Brand
            src={data.logo}
            alt={data.brand}
            style={{ marginTop: '8px' }}
          />
        }
        logoProps={{ href: data.root }}
        showNavToggle
        contextSelector={
          <Toolbar
            data={data}
            currentLocation={currentLocation}
            changeLocation={changeLocation}
            currentOrganization={currentOrganization}
            changeOrganization={changeOrganization}
            isLoading={isLoading}
            changeActiveMenu={changeActiveMenu}
          />
        }
        afterNavToggle={afterNavToggle}
        className="navbar-pf-vertical"
      />
    );
    const sidebar = (
      <PageSidebar
        nav={
          <VerticalNav
            items={items}
            activeMenu={activeMenu}
            changeActiveMenu={changeActiveMenu}
            history={history}
          />
        }
        theme="dark"
      />
    );
    const pageId = 'main';
    const pageSkipToContent = (
      <SkipToContent href={`#${pageId}`}>Skip to content</SkipToContent>
    );

    return (
      <Page
        header={header}
        sidebar={sidebar}
        isManagedSidebar
        skipToContent={pageSkipToContent}
        mainContainerId={pageId}
        className="foreman-theme"
      >
        <PageSection variant={PageSectionVariants.light}>
          {children}
        </PageSection>
      </Page>
    );
  }
}

Layout.propTypes = {
  ...layoutPropTypes,
};

Layout.defaultProps = {
  ...layoutDefaultProps,
};

export default Layout;
