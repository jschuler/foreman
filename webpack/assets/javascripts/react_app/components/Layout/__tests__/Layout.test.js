import React from 'react';
import { mount } from '@theforeman/test';

import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';
import Layout from '../Layout';
import { layoutMock, noItemsMock, hasTaxonomiesMock } from '../Layout.fixtures';

jest.mock('../../notifications', () => 'span');

const didMountStubs = () => ({
  fetchMenuItems: jest.fn(),
  changeLocation: jest.fn(),
  changeOrganization: jest.fn(),
});

const fixtures = {
  'renders layout': layoutMock,
};

describe('Layout', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(Layout, fixtures));

  describe('triggering', () => {
    it('should trigger fetchMenuItems, items array is empty', () => {
      const props = { ...noItemsMock, ...didMountStubs() };
      mount(<Layout {...props} />);

      expect(props.fetchMenuItems.mock.calls).toHaveLength(1);
      expect(props.changeLocation.mock.calls).toHaveLength(1);
      expect(props.changeOrganization.mock.calls).toHaveLength(1);
    });
    it('should not trigger fetchMenuItems, changeLocation, ChangeOrganization', () => {
      const props = { ...hasTaxonomiesMock, ...didMountStubs() };
      const spy = jest.spyOn(console, 'error').mockImplementation();

      mount(<Layout {...props} />);

      // Currently expect a prop warning since VerticalNav passes an object into NavExpandable title prop although it expects a string. This will change soon.
      expect(spy).toHaveBeenCalledTimes(1);
      // eslint-disable-next-line
      expect(console.error.mock.calls[0][0]).toEqual(expect.stringContaining('Warning: Failed prop type: Invalid prop `title` of type `object` supplied to `NavExpandable`, expected `string`'));
      spy.mockRestore();

      expect(props.fetchMenuItems.mock.calls).toHaveLength(0);
      expect(props.changeLocation.mock.calls).toHaveLength(0);
      expect(props.changeOrganization.mock.calls).toHaveLength(0);
    });
  });
});
