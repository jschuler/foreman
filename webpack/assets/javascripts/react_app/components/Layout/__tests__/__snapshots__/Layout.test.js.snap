// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout rendering renders layout 1`] = `
<Page
  breadcrumb={null}
  className="foreman-theme"
  defaultManagedSidebarIsOpen={true}
  header={
    <CustomPageHeader
      afterNavToggle={[Function]}
      aria-label="Global navigation"
      avatar={null}
      className="navbar-pf-vertical"
      contextSelector={
        <Toolbar
          changeActiveMenu={[Function]}
          changeLocation={[Function]}
          changeOrganization={[Function]}
          currentLocation="Any Location"
          currentOrganization="Any Organization"
          data={
            Object {
              "brand": "Foreman",
              "locations": Object {
                "available_locations": Array [
                  Object {
                    "href": "/locations/1-yaml/select",
                    "id": 1,
                    "title": "yaml",
                  },
                  Object {
                    "href": "/locations/2-london/select",
                    "id": 2,
                    "title": "london",
                  },
                  Object {
                    "href": "/locations/3-norway/select",
                    "id": 3,
                    "title": "norway",
                  },
                ],
                "current_location": "london",
              },
              "logo": "/assets/header_logo-c9614c16f2ee399ae9cb7f36ec94b9a26bf8cf9eabaa7fe6099bf80d1f7940db.svg",
              "menu": Array [
                Object {
                  "children": Array [
                    Object {
                      "name": "Dashboard",
                      "type": "item",
                      "url": "/",
                    },
                    Object {
                      "name": "Facts",
                      "type": "item",
                      "url": "/fact_values",
                    },
                  ],
                  "icon": "fa fa-tachometer",
                  "name": "Monitor",
                  "type": "sub_menu",
                },
                Object {
                  "children": Array [
                    Object {
                      "name": "All Hosts",
                      "type": "item",
                      "url": "/hosts/new",
                    },
                    Object {
                      "name": "Architectures",
                      "type": "item",
                      "url": "/architectures",
                    },
                  ],
                  "icon": "fa fa-server",
                  "name": "Hosts",
                  "type": "sub_menu",
                },
                Object {
                  "children": Array [
                    Object {
                      "name": "Environments",
                      "type": "item",
                      "url": "/environments",
                    },
                    Object {
                      "name": "Architectures",
                      "type": "item",
                      "url": "/architectures",
                    },
                  ],
                  "icon": "fa fa-wrench",
                  "name": "User",
                  "type": "sub_menu",
                },
                Object {
                  "children": Array [
                    Object {
                      "name": "Domains",
                      "type": "item",
                      "url": "/domains",
                    },
                    Object {
                      "name": "Realms",
                      "type": "item",
                      "url": "/realms",
                    },
                  ],
                  "icon": "pficon pficon-network",
                  "name": "Infrastructure",
                  "type": "sub_menu",
                },
              ],
              "notification_url": "/notification_recipients",
              "orgs": Object {
                "available_organizations": Array [
                  Object {
                    "href": "/organizations/1-org1/select",
                    "id": 1,
                    "title": "org1",
                  },
                  Object {
                    "href": "/organizations/2-org2/select",
                    "id": 2,
                    "title": "org2",
                  },
                ],
                "current_org": "org1",
                "many_organizations": Array [
                  Object {
                    "href": "/organizations/1-org1/select",
                    "id": 1,
                    "title": "org1",
                  },
                  Object {
                    "href": "/organizations/2-org2/select",
                    "id": 2,
                    "title": "org2",
                  },
                  Object {
                    "href": "/organizations/3-org3/select",
                    "id": 3,
                    "title": "org3",
                  },
                  Object {
                    "href": "/organizations/4-org4/select",
                    "id": 4,
                    "title": "org4",
                  },
                  Object {
                    "href": "/organizations/5-org5/select",
                    "id": 5,
                    "title": "org5",
                  },
                  Object {
                    "href": "/organizations/6-org6/select",
                    "id": 6,
                    "title": "org6",
                  },
                  Object {
                    "href": "/organizations/7-org7/select",
                    "id": 7,
                    "title": "org7",
                  },
                ],
              },
              "root": "/",
              "stop_impersonation_url": "/users/stop_impersonation",
              "taxonomies": Object {
                "locations": true,
                "organizations": true,
              },
              "user": Object {
                "current_user": Object {
                  "user": Object {
                    "firstname": "Admin",
                    "id": 4,
                    "lastname": "User",
                    "login": "admin",
                    "name": "Admin User",
                  },
                },
                "user_dropdown": Array [
                  Object {
                    "children": Array [
                      Object {
                        "href": "/Aa",
                        "isDivider": false,
                        "onClick": [MockFunction],
                        "title": "Aa",
                      },
                      Object {
                        "href": "/Cc",
                        "isDivider": false,
                        "onClick": [MockFunction],
                        "title": "Cc",
                      },
                    ],
                    "icon": "fa fa-user",
                    "name": "User",
                    "type": "sub_menu",
                  },
                ],
              },
            }
          }
          isLoading={false}
        />
      }
      isNavOpen={true}
      logo={
        <Brand
          alt="Foreman"
          src="/assets/header_logo-c9614c16f2ee399ae9cb7f36ec94b9a26bf8cf9eabaa7fe6099bf80d1f7940db.svg"
          style={
            Object {
              "marginTop": "8px",
            }
          }
        />
      }
      logoComponent="a"
      logoProps={
        Object {
          "href": "/",
        }
      }
      onNavToggle={[Function]}
      showNavToggle={true}
      toolbar={null}
      topNav={null}
    />
  }
  isManagedSidebar={true}
  mainContainerId="main"
  onPageResize={[Function]}
  sidebar={
    <PageSidebar
      nav={
        <VerticalNav
          activeMenu="Monitor"
          changeActiveMenu={[Function]}
          history={
            Object {
              "push": [MockFunction],
            }
          }
          items={
            Array [
              Object {
                "href": "/a",
                "iconClass": "fa fa-tachometer",
                "initialActive": true,
                "subItems": Array [
                  Object {
                    "href": "/Aa",
                    "isDivider": false,
                    "onClick": [MockFunction],
                    "title": "Aa",
                  },
                  Object {
                    "href": "/Cc",
                    "isDivider": false,
                    "onClick": [MockFunction],
                    "title": "Cc",
                  },
                ],
                "title": "Monitor",
              },
              Object {
                "href": "/b",
                "iconClass": "fa fa-server",
                "initialActive": false,
                "subItems": Array [
                  Object {
                    "href": "/Dd",
                    "isDivider": false,
                    "onClick": [MockFunction],
                    "title": "Dd",
                  },
                  Object {
                    "href": "/Ff",
                    "isDivider": false,
                    "onClick": [MockFunction],
                    "title": "Ff",
                  },
                ],
                "title": "Hosts",
              },
              Object {
                "href": "/c",
                "iconClass": "fa fa-wrench",
                "initialActive": false,
                "subItems": Array [
                  Object {
                    "href": "/Gg",
                    "isDivider": false,
                    "onClick": [MockFunction],
                    "title": "Gg",
                  },
                  Object {
                    "href": "/Ii",
                    "isDivider": false,
                    "onClick": [MockFunction],
                    "title": "Ii",
                  },
                ],
                "title": "Configure",
              },
            ]
          }
        />
      }
      theme="dark"
    />
  }
  skipToContent={
    <SkipToContent
      className=""
      component="a"
      href="#main"
      show={false}
    >
      Skip to content
    </SkipToContent>
  }
>
  <PageSection
    variant="light"
  />
</Page>
`;
